### PR TITLE
QUIC: Demo-Driven Design Demos Update & Report

### PR DESCRIPTION
**Pull request by hlandau**
This updates the QUIC DDD demos which were produced as part of the API design process to minimise the API changes which would be needed for existing applications.

These updated demos are all tested and working with QUIC.

Attention is drawn to the following artefacts:

- (Old) [The original README for the DDD demos.](https://github.com/hlandau/openssl/blob/quic-ddd-update-2/doc/designs/ddd/README.md)
- (New, recommended reading) [**The final report with conclusions on the DDD process.**](https://github.com/hlandau/openssl/blob/quic-ddd-update-2/doc/designs/ddd/REPORT.md) The report summarises the planned and unplanned changes to each demo.
- The originally planned, and actually realised changes, for each demo. The planned changes were originally formulated as diffs, however for the sake of readability and maintenance, they are now represented here as guarded `#ifdef USE_QUIC` code sections:

  <table><tr><th>Demo</th><th>Planned Changes</th><th>Unplanned Changes</th></tr>
  <tr><td>ddd-01-conn-blocking</td><td><a href="https://github.com/openssl/openssl/pull/21715/commits/54251edbf90524caff942b0c35a17cf2c4abc7ac">commit</a></td><td><a href="https://github.com/openssl/openssl/pull/21715/commits/cd44ab37594c0df25c68c135bfbb7d8bed644caa">commit</a></td></tr>
  <tr><td>ddd-02-conn-nonblocking</td><td><a href="https://github.com/openssl/openssl/pull/21715/commits/b27ce91f2decfe9b01125bf8103be245467201b4">commit</a></td><td><a href="https://github.com/openssl/openssl/pull/21715/commits/5e0e97c3d76b3058628b330d29576e0ab6cd2eb1">commit</a></td></tr>
  <tr><td>ddd-03-fd-blocking</td><td><a href="https://github.com/openssl/openssl/pull/21715/commits/8b5183d62fae3784790594353d53e27628b9b04f">commit</a></td><td><a href="https://github.com/openssl/openssl/pull/21715/commits/13041ec59b807075f788fb50a67954e1aaba55da">commit</a></td></tr>
  <tr><td>ddd-04-fd-nonblocking</td><td><a href="https://github.com/openssl/openssl/pull/21715/commits/dd856c23ea77120283fa1a5a3154851a6435b06a">commit</a></td><td><a href="https://github.com/openssl/openssl/pull/21715/commits/9f5e1f3841e3ae39ca226b77da05baf339e6a5f2">commit</a></td></tr>
  <tr><td>ddd-05-mem-nonblocking</td><td><a href="https://github.com/openssl/openssl/pull/21715/commits/95d34a9c15acd44ede08adf4bfa52c43fe5b0871">commit</a></td><td><a href="https://github.com/openssl/openssl/pull/21715/commits/cac62cead88db498555d90332ae14144c57d4076">commit</a></td></tr>
  <tr><td>ddd-06-mem-uv</td><td><a href="https://github.com/openssl/openssl/pull/21715/commits/efa409741b4f606e32120033cb3542ccf0d4ff9d">commit</a></td><td><a href="https://github.com/openssl/openssl/pull/21715/commits/acda8fdba301088c86899a3fa6b39b25f1431832">commit</a></td></tr>
  </table>

This PR also incorporates changes to the QUIC code which were overlooked until now but required to get the DDD demos working as planned. In particular:

- BIO_s_connect adapted to support datagram mode
- Make BIO_ssl helper functions configure BIO_s_connect for QUIC automatically when QUIC is being used
- Automatically configure network-side BIOs as nonblocking
- Refactor of can-poll flags out of QUIC_CHANNEL and into QUIC_REACTOR.
- Refactor of QUIC APL blocking configuration to allow our determination of whether we can support blocking mode to be made as late as possible.
- Introduce concept of *addressed mode* v. *non-addressed mode* in QUIC_CHANNEL. A non-addressed QUIC connection is one used with e.g. a BIO_s_dgram_pair which is not providing L4 source/destination addresses. This kind of connection will not be able to support connection migration, multipath, etc. in the future.
- Add `BIO_dgram_detect_peer_addr` API to avoid overloading the meaning of `BIO_dgram_get_peer`.
- Wired BIO_sendmmsg/BIO_recvmmsg capability negotiation for BIO_s_datagram.

```
=== Report
f130d2d098 QUIC DDD: Final report

=== Unplanned/planned changes
4cc7decd5d QUIC DDD: ddd-02-conn-nonblocking-threads: Unplanned changes
adbf5d843d QUIC DDD: ddd-02-conn-nonblocking-threads: Planned changes
acda8fdba3 QUIC DDD: ddd-06-mem-uv: Unplanned changes
efa409741b QUIC DDD: ddd-06-mem-uv: Planned changes
cac62cead8 QUIC DDD: ddd-05-mem-nonblocking: Unplanned changes
95d34a9c15 QUIC DDD: ddd-05-mem-nonblocking: Planned changes
9f5e1f3841 QUIC DDD: ddd-04-fd-nonblocking: Unplanned changes
dd856c23ea QUIC DDD: ddd-04-fd-nonblocking: Planned changes
13041ec59b QUIC DDD: ddd-03-fd-blocking: Unplanned changes
8b5183d62f QUIC DDD: ddd-03-fd-blocking: Planned changes
5e0e97c3d7 QUIC DDD: ddd-02-conn-nonblocking: Unplanned changes
b27ce91f2d QUIC DDD: ddd-02-conn-nonblocking: Planned changes
cd44ab3759 QUIC DDD: ddd-01-conn-blocking: Unplanned changes
54251edbf9 QUIC DDD: ddd-01-conn-blocking: Planned changes

=== Minor rework of DDD demos
7f013fa310 QUIC DDD: Update makefile
b9ad93c1bc QUIC DDD: Add unchanged copy of ddd-02-conn-nonblocking to serve as base for thread-assisted variant
ea7c74166e QUIC DDD: Allow target host:port to be set from command line
226dc1326c QUIC DDD: Fix bug in ddd-06-mem-uv

=== Supporting changes
c259499e10 QUIC APL: Introduce addressed v. non-addressed mode handling
c84b251473 BIO_s_dgram_pair: Correct implementation of BIO_CTRL_DGRAM_GET_LOCAL_ADDR_ENABLE
8fddce61f3 BIO_s_datagram: Wire capability negotiation for BIO_s_datagram
96e9f81bcb BIO: Add BIO_dgram_detect_peer_addr API
fee36353f0 BIO_s_datagram: Support configuring non-blocking mode
9d18ee041d BIO_s_connect: Support configuration of non-blocking mode in datagram mode
63ed11bfa1 QUIC CHANNEL: Introduce concept of (non-)addressed mode
1f04b6d32e QUIC APL: Refactor blocking configuration to allow late blocking support detection
d71a732f72 QUIC CHANNEL: Cleanup poll descriptor management
5323505d45 QUIC REACTOR: Move can-poll flags into reactor
c00ff9ff7b QUIC APL: Autoconfigure BIOs as non-blocking
231bc4e419 BIO_ssl: Make helper functions configure BIOs for QUIC correctly
139fda5cf2 QUIC APL: Better error reporting
099068e717 QUIC CHANNEL: Only handle the first protocol error raised
1d6b5674d4 BIO_s_connect: Add support for datagram mode
277fb45d17 BIO_s_connect: Make internal functions static
e3a2f0d832 QUIC TLS: Better error message when ALPN not used
```

Fixes https://github.com/openssl/project/issues/40

### CODE REVIEW

**Comment by hlandau** [comment]
Forgot the manpages for the added BIO APIs. Forthcoming.

File: doc/man3/BIO_s_connect.pod
Comment (hlandau) [reply doc/man3/BIO_s_connect.pod:116]: It doesn't &mdash; clarified.

File: doc/designs/ddd/ddd-02-conn-nonblocking.c
Comment (hlandau) [reply doc/designs/ddd/ddd-02-conn-nonblocking.c:91]: It's an emergent consequence of the existing design. The push ctrl is not handled by QUIC but BIO_push does not really handle failure right. So the buffering BIO is basically just hidden on the BIO stack but not used until the chain is freed.

**Comment by mattcaswell** [comment]
Ping - needs second review.

@hlandau - needs rebase too

**Comment by hlandau** [comment]
This now depends on #21829

File: crypto/bio/bss_dgram_pair.c
Comment (t8m) [inline crypto/bio/bss_dgram_pair.c:696]: Why cast to long and then assign to int?

File: doc/designs/ddd/ddd-05-mem-nonblocking.c
Comment (t8m) [inline doc/designs/ddd/ddd-05-mem-nonblocking.c:196]: s/frame/datagram/?

File: doc/designs/ddd/ddd-05-mem-nonblocking.c
Comment (t8m) [inline doc/designs/ddd/ddd-05-mem-nonblocking.c:207]: As above?

File: doc/designs/ddd/REPORT.md
Comment (t8m) [inline doc/designs/ddd/REPORT.md:154]: Typo: substantially

**Review by t8m — CHANGES_REQUESTED** [review]
Good work! Just minor things.

**Comment by hlandau** [comment]
Updated and rebased.

**Review by t8m — APPROVED** [review]
Good work!

**Comment by t8m** [comment]
@mattcaswell please reconfirm

**Review by mattcaswell — APPROVED** [review]

**Pull request closed without merging**
